urld558a3d988_gao715108023_resourceMonitor_tgz-pJ8-com_sinopec_agent_AgentJ8/src/com/sinopec/group/rpc/TopologyClient.java

```diff
@@ -0,0 +1,122 @@
+package com.sinopec.group.rpc;
+
+import java.net.MalformedURLException;
+import java.rmi.Naming;
+import java.rmi.NotBoundException;
+import java.rmi.Remote;
+import java.rmi.RemoteException;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import com.sinopec.bean.AdjacentGroup;
+import com.sinopec.dao.GroupInfoDao;
+import com.sinopec.io.GroupStatus;
+import com.sinopec.utils.LogUtils;
+
+public class TopologyClient implements Runnable {
+
+	private static final Log LOG = LogFactory.getLog(TopologyClient.class);
+
+	private Remote remote;
+
+	private static volatile boolean running = true;
+
+	private static volatile boolean service = true;
+
+	private String crontrolNodeIP;
+
+	private int port;
+
+	private int samplingInterval = 3000;
+
+	private String localhostIP;
+
+	private AdjacentGroup adjacentGroup;
+
+	public TopologyClient(String crontrolNodeIP, int port, int samplingInterval, String localhostIP, AdjacentGroup adjacentGroup) {
+		super();
+		this.crontrolNodeIP = crontrolNodeIP;
+		this.port = port;
+		this.samplingInterval = samplingInterval;
+		this.localhostIP = localhostIP;
+		this.adjacentGroup = adjacentGroup;
+	}
+
+	public boolean isServerExists(String bindName) {
+		boolean result = false;
+		try {
+			remote = Naming.lookup(bindName);
+			if (remote != null)
+				result = true;
+		} catch (MalformedURLException e) {
+			// TODO Auto-generated catch block
+			e.printStackTrace();
+		} catch (RemoteException e) {
+			// TODO Auto-generated catch block
+			e.printStackTrace();
+		} catch (NotBoundException e) {
+			// TODO Auto-generated catch block
+			e.printStackTrace();
+		}
+		return result;
+	}
+
+	public void connectToCrontrolNode() {
+		int retryCount = 1;
+		while (running) {
+			if (isServerExists("rmi://" + crontrolNodeIP + ":" + port + "/topology")) {
+				running = false;
+				service = true;
+			} else {
+				LOG.info("Retrying connect to CrontrolNode: " + crontrolNodeIP + ".Already tries " + retryCount + " time(s).");
+				retryCount++;
+				try {
+					Thread.sleep(3000);
+				} catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)